domain/secretbackend/state/operation.go

```diff
@@ -0,0 +1,208 @@
+		// New insert.
+		if o.Name == "" {
+			return fmt.Errorf("%w: name is missing", backenderrors.NotValid)
+		}
+		if o.BackendType == "" {
+			return fmt.Errorf("%w: type is missing", backenderrors.NotValid)
+		}
+		return nil
+	} else if err != nil {
+		return errors.Trace(err)
+	}
+	// Update.
+	if existing.BackendType == "" {
+		// This should never happen, but just in case.
+		return errors.Errorf("backend type is empty for backend %q", o.ID)
+	}
+	if existing.Name == "" {
+		// This should never happen, but just in case.
+		return errors.Errorf("backend name is empty for backend %q", o.ID)
+	}
+	if o.BackendType != "" && o.BackendType != existing.BackendType {
+		// The secret backend type is immutable.
+		return errors.Errorf(
+			"cannot change backend type from %q to %q because backend type is immutable",
+			existing.BackendType, o.BackendType,
+		)
+	}
+	// Fill in the existing backend type.
+	o.BackendType = existing.BackendType
+	if o.Name == "" {
+		// Fill in the existing name.
+		o.Name = existing.Name
+	}
+	if o.TokenRotateInterval == nil && existing.TokenRotateInterval.Valid {
+		// Fill in the existing token rotate interval.
+		o.TokenRotateInterval = &existing.TokenRotateInterval.Duration
+	}
+	return nil
+}
+
+// apply applies the upsert operation to the database.
+func (o *upsertOperation) apply(ctx context.Context, tx *sqlair.TX) error {
+	if err := o.prepareData(ctx, tx); err != nil {
+		return errors.Trace(err)
+	}
+
+	sb := SecretBackend{
+		ID:          o.ID,
+		Name:        o.Name,
+		BackendType: o.BackendType,
+	}
+	if o.TokenRotateInterval != nil {
+		sb.TokenRotateInterval = domain.NullableDuration{Duration: *o.TokenRotateInterval, Valid: true}
+	}
+
+	err := tx.Query(ctx, o.upsertBackendStmt, sb).Run()
+	if database.IsErrConstraintUnique(err) {
+		return fmt.Errorf("%w: name %q", backenderrors.AlreadyExists, sb.Name)
+	}
+	if err != nil {
+		return fmt.Errorf("cannot upsert secret backend %q: %w", o.Name, err)
+	}
+	if o.NextRotateTime != nil && !o.NextRotateTime.IsZero() {
+		err = tx.Query(ctx, o.upsertRotationStmt, SecretBackendRotation{
+			ID:               o.ID,
+			NextRotationTime: sql.NullTime{Time: *o.NextRotateTime, Valid: true},
+		}).Run()
+		if err != nil {
+			return fmt.Errorf("cannot upsert secret backend rotation time for %q: %w", o.Name, err)
+		}
+	}
+	if len(o.Config) == 0 {
+		return nil
+	}
+	namesToKeep := make(sqlair.S, 0, len(o.Config))
+	for k := range o.Config {
+		namesToKeep = append(namesToKeep, k)
+	}
+	err = tx.Query(ctx, o.clearConfigStmt, sqlair.M{"uuid": o.ID}, namesToKeep).Run()
+	if err != nil {
+		return fmt.Errorf("cannot clear secret backend config for %q: %w", o.Name, err)
+	}
+	for k, v := range o.Config {
+		err = tx.Query(ctx, o.upsertConfigStmt, SecretBackendConfig{
+			ID:      o.ID,
+			Name:    k,
+			Content: v.(string),
+		}).Run()
+		if err != nil {
+			return fmt.Errorf("cannot upsert secret backend config for %q: %w", o.Name, err)
+		}
+	}
```

Comment:
How big can `o.Config` be in terms of size? Depending on the size, you could end up stalling dqlite read queries. I would suggest maybe rethinking this.

Reply:
Yeah, I did think about this. The config usually just has a few fields but we don't have a limit on it.
But the sqlair does not have bulk support yet, probably we should just use raw SQL for now or leave it as a TODO.